data-model/src/main/kotlin/io/micronaut/data/repository/kotlin/CoroutineCrudRepository.kt

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2017-2021 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.repository.kotlin
+
+import io.micronaut.data.repository.GenericRepository
+import kotlinx.coroutines.flow.Flow
+
+/**
+ * Interface for CRUD repository using Kotlin coroutines.
+ *
+ * @param <E> The entity type
+ * @param <ID> The ID type
+ *
+ * @author Denis Stepanov
+ * @since 3.1.0
+ */
+interface CoroutineCrudRepository<E, ID> : GenericRepository<E, ID> {
+
+    /**
+     * Saves the given valid entity, returning a possibly new entity representing the saved state. Note that certain implementations may not be able to detect whether a save or update should be performed and may always perform an insert. The [.update] method can be used in this case to explicitly request an update.
+     *
+     * @param entity The entity to save. Must not be null.
+     * @return The saved entity will never be null.
+     * @param <S> The generic type
+     */
+    suspend fun <S : E> save(entity: S): S
+
+    /**
+     * This method issues an explicit update for the given entity. The method differs from [.save] in that an update will be generated regardless if the entity has been saved previously or not. If the entity has no assigned ID then an exception will be thrown.
+     *
+     * @param entity The entity to save. Must not be null.
+     * @return The updated entity will never be null.
+     * @param <S> The generic type
+     */
+    suspend fun <S : E> update(entity: S): S
+
+    /**
+     * This method issues an explicit update for the given entities. The method differs from [.saveAll] in that an update will be generated regardless if the entity has been saved previously or not. If the entity has no assigned ID then an exception will be thrown.
+     *
+     * @param entities The entities to update. Must not be null.
+     * @return The updated entities will never be null.
+     * @param <S> The generic type
+     */
+    fun <S : E> updateAll(entities: Iterable<S>): Flow<S>
```

Comment:
I wonder if this should be `suspend` and return a `Iterable` and then there should be a separate `FlowCrudRepository`. Seems we are mixing the two APIs here?

Reply:
I think this is the correct way and it's the same as in spring-data, calling `toList` on `Flow` will give you the collection.